internal/pkg/addon/addons.go

```diff
@@ -33,49 +39,115 @@ var (
 type workspaceReader interface {
 	WorkloadAddonsPath(name string) string
 	WorkloadAddonFilePath(wkldName, fName string) string
+	EnvAddonsPath() string
+	EnvAddonFilePath(fName string) string
 	ListFiles(dirPath string) ([]string, error)
 	ReadFile(fPath string) ([]byte, error)
 }
 
-// Stack represents a CloudFormation stack.
-type Stack struct {
-	template     *cfnTemplate
-	parameters   yaml.Node
+// WorkloadAddonStack represents a CloudFormation stack for workload addons.
+type WorkloadAddonStack struct {
+	stack
 	workloadName string
 }
 
+type EnvironmentAddonStack struct {
+	stack
+}
+
+type stack struct {
+	template   *cfnTemplate
+	parameters yaml.Node
+}
+
+type parser struct {
+	ws                 workspaceReader
+	filePath           func(fName string) string
+	validateParameters func(params yaml.Node, paramFilePath string) error
+}
+
 // Parse parses the 'addon/' directory for the given workload
 // and returns a Stack created by merging the CloudFormation templates
 // files found there. If no addons are found, Parse returns a nil
 // Stack and ErrAddonsNotFound.
-func Parse(workloadName string, ws workspaceReader) (*Stack, error) {
+func Parse(workloadName string, ws workspaceReader) (*WorkloadAddonStack, error) {
 	fNames, err := ws.ListFiles(ws.WorkloadAddonsPath(workloadName))
 	if err != nil {
-		return nil, &ErrAddonsNotFound{
-			WlName:    workloadName,
+		return nil, fmt.Errorf("list addons for workload %s: %w", workloadName, &ErrAddonsNotFound{
 			ParentErr: err,
-		}
+		})
 	}
 
-	template, err := parseTemplate(fNames, workloadName, ws)
+	parser := parser{
+		ws: ws,
+		filePath: func(fName string) string {
+			return ws.WorkloadAddonFilePath(workloadName, fName)
+		},
+		validateParameters: func(params yaml.Node, paramFilePath string) error {
+			return validateReservedParameters(params, wkldAddonsParameterReservedKeys)
+		},
+	}
+
+	template, err := parser.parseTemplate(fNames)
 	if err != nil {
 		return nil, err
 	}
 
-	params, err := parseParameters(fNames, workloadName, ws)
+	params, err := parser.parseParameters(fNames)
 	if err != nil {
 		return nil, err
 	}
 
-	return &Stack{
-		template:     template,
-		parameters:   params,
+	return &WorkloadAddonStack{
+		stack: stack{
+			template:   template,
+			parameters: params,
+		},
 		workloadName: workloadName,
 	}, nil
 }
 
+// ParseEnvAddon parses the 'addon/' directory for environments
+// and returns a Stack created by merging the CloudFormation templates
+// files found there. If no addons are found, Parse returns a nil
+// Stack and ErrAddonsNotFound.
+func ParseEnvAddon(ws workspaceReader) (*EnvironmentAddonStack, error) {
+	fNames, err := ws.ListFiles(ws.EnvAddonsPath())
+	if err != nil {
+		return nil, fmt.Errorf("list addons for environments: %w", &ErrAddonsNotFound{
+			ParentErr: err,
+		})
+	}
+	parser := parser{
+		ws: ws,
+		filePath: func(fName string) string {
+			return ws.EnvAddonFilePath(fName)
+		},
```

Comment:
nit: would this work?
```suggestion
		filePath: ws.EnvAddonFilePath,
```